File: Medium/lengthOfLongestSubstring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s) == 1:
            return 1
        lead, follow = 0, 0
        mySet = {}
        max = 0
        while lead < len(s):
            if s[lead] not in mySet:
                mySet[s[lead]] = lead
                lead += 1
                if lead - follow > max:
                    max = lead - follow
            else:
                if lead - follow > max:
                    max = lead - follow
                follow, lead = mySet[s[lead]], mySet[s[lead]] + 1
                mySet = {}
        return max




if __name__ == '__main__':
    solution = Solution()

    # testCase = "aud"
    # ans = solution.lengthOfLongestSubstring(testCase)
    # print(f"{ans}")

    testCase = "fasdgbttttthjkloiuytredf"
    ans = solution.lengthOfLongestSubstring(testCase)
    print(f"{ans}")

File: cppWorkingDirectory/.vscode/lengthOfLongestSubstring.cpp
#include <string>
#include <iostream>
#include <set>
#include <algorithm>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        int ans = 0;
        int l = 0;
        int r = 0;
        std::set<char> mySet;
        while (r < s.length()) {
            if (mySet.find(s[r]) != mySet.end()) {
                while (s[l] != s[r]) {
                    mySet.erase(s[l]);
                    l++;
                }
                l++;
            } else {
                mySet.insert(s[r]);
            }
            ans = std::max(ans, r - l + 1);
            r++;
        }
        return ans;
    }
};


int main () {
    Solution solution;
    std::string s = "abcabcbb";
    int ans = solution.lengthOfLongestSubstring(s);
    std::cout << ans << std::endl;
}